src/main/java/com/example/reservation/reservation/service/ReservationService.java

```diff
@@ -0,0 +1,101 @@
+import com.example.reservation.reservation.domain.Reservation;
+import com.example.reservation.reservation.domain.ReservationStatus;
+import com.example.reservation.reservation.dto.ReservationRequestDto;
+import com.example.reservation.reservation.domain.ReservationRepository;
+import com.example.reservation.reservation.dto.ReservationResponseDto;
+import com.example.reservation.room.domain.Room;
+import com.example.reservation.room.repostitory.RoomRepository;
+import com.example.reservation.user.domain.User;
+import com.example.reservation.user.domain.UserRepository;
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ReservationService {
+
+  private final UserRepository userRepository;
+  private final RoomRepository roomRepository;
+  private final ReservationRepository reservationRepository;
+  private final PaymentRepository paymentRepository;
+
+  @Transactional
+  public ReservationResponseDto createReservation(ReservationRequestDto request) {
+    User user = userRepository.findById(request.getUserId())
+        .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 사용자입니다."));
+
+    Room room = roomRepository.findById(request.getRoomId())
+        .orElseThrow(() -> new IllegalArgumentException("해당 방이 존재하지 않습니다."));
+
+    Accommodation accommodation = room.getAccommodation();
+
+    validateReservationDate(room.getId(), request.getCheckinDate(), request.getCheckoutDate());
+
+    int totalPrice = calculateTotalPrice(room, request.getCheckinDate(), request.getCheckoutDate());
+
+    Reservation reservation = reservationRepository.save(
+        Reservation.builder()
+            .user(user)
+            .room(room)
+            .checkinDate(request.getCheckinDate())
+            .checkoutDate(request.getCheckoutDate())
+            .guestCount(request.getGuestCount())
+            .totalPrice(totalPrice)
+            .status(ReservationStatus.RESERVED)
+            .build()
+    );
+
+    // 결제 정보 저장
+    paymentRepository.save(Payment.builder()
+        .reservation(reservation)
+        .amount(totalPrice)
+        .state(PaymentState.PAID)
+        .paymentAt(LocalDateTime.now())
+        .build());
+
+
+    return ReservationResponseDto.from(
+        reservation,
+        accommodation.getCheckinTime(),
+        accommodation.getCheckoutTime()
+    );
+  }
+
+  private int calculateTotalPrice(Room room, LocalDate checkin, LocalDate checkout) {
+    int totalPrice = 0;
+
+    while (checkin.isBefore(checkout)) {
+      DayOfWeek day = checkin.getDayOfWeek();
+      boolean isWeekend = (day == DayOfWeek.FRIDAY || day == DayOfWeek.SATURDAY); // 금,토 주말요금
+
+      totalPrice += isWeekend ? room.getPriceWeekend() : room.getPriceWeekday();
+      checkin = checkin.plusDays(1);
+    }
```

Comment:
while문은 가능하면 피하는게 좋습니다. 혹시 휴먼에러로 while 조건문이 깨질경우 무한루프에 빠질수가 있어서요

```suggestion
            Map<PriceType, Long> result = checkin.datesUntil(checkout).collect(Collectors.groupingBy(Main::getPrice, Collectors.counting()));
           int price = room.getPriceWeekend() * result.get(PriceType.WEEKEND) + room.getPriceWeek() * result.get(PriceType.WEEK)
    /// 메소드 밖에서
    private PriceType getPrice(LocalDate date) {
        if (date.getDayOfWeek() == DayOfWeek.FRIDAY || date.getDayOfWeek() == DayOfWeek.SATURDAY) {
            return PriceType.WEEKEND;
        } else {
            return PriceType.WEEK;
        }
    }

    enum PriceType {
        WEEK,
        WEEKEND
    }

```

Reply:
@mingming-mentor  와..  Map으로 처리하는 건 생각도 못했습니다! 수정 했습니다!